### PR TITLE
Code clean up 

### PR DESCRIPTION
**Pull request by sunank200**
- Refactor and reduce code complexity
- Code clean-up
- Add exception handling 
- Use the Sanic framework for configurations
- Update doc-strings and typing

closes: #35

### CODE REVIEW

File: api/ask_astro/rest/controllers/get_request.py [hunk 1]
```diff
@@ -15,15 +16,23 @@
...
     Handles GET requests to the /requests/{request_id} endpoint.
+
+    :param request: The Sanic request object.
+    :param request_id: The unique identifier for the AskAstro request.
     """
-    logger.info("Received GET request for request %s", request_id)
-    request = await firestore_client.collection(FirestoreCollections.requests).document(str(request_id)).get()
-    logger.info("Request %s exists: %s", request_id, request.exists)
+    try:
+        logger.info("Received GET request for request %s", request_id)
```
Comment (Lee-W) [inline api/ask_astro/rest/controllers/get_request.py:27]: Should we consolidate how we format string in logger?

File: api/ask_astro/clients/weaviate_.py [hunk 2]
```diff
@@ -1,22 +1,30 @@
...
 from ask_astro.config import AzureOpenAIParams, WeaviateConfig
 from langchain.embeddings import OpenAIEmbeddings
 from langchain.vectorstores import Weaviate
 from weaviate import Client as WeaviateClient
 
+# Initialize OpenAI embeddings using the specified parameters.
 embeddings = OpenAIEmbeddings(
     **AzureOpenAIParams.us_east,
     deployment="text-embedding-ada-002",
     model="text-embedding-ada-002",
```
[the remark this answers is not in the document]
Comment (sunank200) [reply api/ask_astro/clients/weaviate_.py]: https://github.com/astronomer/ask-astro/pull/61/commits/6fcb98302d73292291893ccf73e598f511379272

File: api/ask_astro/rest/controllers/get_request.py (hunk 1 above)
Comment (sunank200) [reply api/ask_astro/rest/controllers/get_request.py:27]: changed it: https://github.com/astronomer/ask-astro/pull/61/commits/b046f9131f44c5fd7026927fd2abe53a6379cda8

File: api/ask_astro/config.py [hunk 3]
```diff
@@ -1,53 +1,59 @@
...
 
 class WeaviateConfig:
-    OpenAIApiKey = os.environ["OPENAI_API_KEY"]
-    url = os.environ["WEAVIATE_URL"]
-    api_key = os.environ["WEAVIATE_API_KEY"]
-    index_name = os.environ["WEAVIATE_INDEX_NAME"]
-    text_key = os.environ["WEAVIATE_TEXT_KEY"]
+    """Contains the config variables for the Weaviate API."""
+
+    OpenAIApiKey = os.environ.get("OPENAI_API_KEY")
```
Comment (Lee-W) [inline api/ask_astro/config.py:54]: May I know why this variable needs to be in camel case? Is it required by open AI?

File: api/ask_astro/rest/controllers/list_recent_requests.py [hunk 4]
```diff
@@ -25,23 +33,21 @@ def to_dict(self):
...
 
-    return json(
-        RecentRequestsResponse(
-            requests=[AskAstroRequest.from_dict(request.to_dict()) for request in requests]
-        ).to_dict(),
-        status=200,
-    )
+        return json(RecentRequestsResponse(requests=recent_requests).to_dict(), status=200)
+    except Exception as e:
+        logger.error("Error while fetching recent requests: %s", str(e))
```
Comment (Lee-W) [inline api/ask_astro/rest/controllers/list_recent_requests.py:52]:
```suggestion
        logger.error("Error while fetching recent requests: %s", e)
```

File: api/ask_astro/rest/controllers/submit_feedback.py [hunk 5]
```diff
@@ -1,28 +1,41 @@
...
+    try:
+        positive = request.json["positive"]
 
-    await submit_feedback(str(request_id), positive, {"source": "api"})
+        await submit_feedback(str(request_id), positive, {"source": "api"})
 
-    return HTTPResponse(status=200)
+        return HTTPResponse(status=200)
+    except Exception as e:
+        logger.error("Error occurred while processing feedback for request %s: %s", request_id, str(e))
```
Comment (Lee-W) [inline api/ask_astro/rest/controllers/submit_feedback.py:40]:
```suggestion
        logger.error("Error occurred while processing feedback for request %s: %s", request_id, e)
```

File: api/ask_astro/services/feedback.py [hunk 6]
```diff
@@ -10,37 +12,50 @@
...
+                        key="correctness",
+                        run_id=langchain_run_id,
+                        score=1 if correct else 0,
+                        source_info=source_info,
+                    )
                 )
             )
-        )
+    except Exception as e:
+        logger.error("Error occurred while processing feedback for request %s: %s", request_id, str(e))
```
Comment (Lee-W) [inline api/ask_astro/services/feedback.py:60]:
```suggestion
        logger.error("Error occurred while processing feedback for request %s: %s", request_id, e)
```

File: api/ask_astro/slack/controllers/feedback/bad.py [hunk 7]
```diff
@@ -10,54 +12,81 @@
...
+
+async def _send_response(details: dict[str, str], respond: AsyncRespond) -> None:
+    """
+    Send a response back to the user in Slack.
+
+    :param details: The details extracted from the Slack body.
+    :param respond: Response object from slack_bolt.
+    """
+    await respond(
+        "😥 Thank you for your feedback, <@%s>!" % details["user"],
```
Comment (Lee-W) [inline api/ask_astro/slack/controllers/feedback/bad.py:67]:
```suggestion
        "😥 Thank you for your feedback, <@{details['user']}>!",
```

This is not logging. I think we're good to use f-string

File: api/ask_astro/slack/controllers/feedback/bad.py [hunk 8]
```diff
@@ -10,54 +12,81 @@
...
+    except SlackApiError as e:
+        # ignore the error if the reaction already exists
+        if e.response["error"] != "already_reacted":
+            raise e
+
+    try:
+        await client.reactions_remove(name="thumbsup", channel=details["channel"], timestamp=details["message_ts"])
+    except Exception as e:
+        logger.debug("Failed to remove thumbsup reaction: %s", e)
+        pass
```
Comment (Lee-W) [inline api/ask_astro/slack/controllers/feedback/bad.py:92]:
```suggestion
```

Not sure whether we need this `pass`

File: api/ask_astro/slack/controllers/feedback/good.py [hunk 9]
```diff
@@ -10,54 +12,81 @@
...
+
+async def _send_response(details: dict[str, str], respond: AsyncRespond) -> None:
+    """
+    Send a positive response back to the user in Slack.
+
+    :param details: The details extracted from the Slack body.
+    :param respond: Response object from slack_bolt.
+    """
+    await respond(
+        "☺️ Thank you for your feedback, <@%s>!" % details["user"],
```
Comment (Lee-W) [inline api/ask_astro/slack/controllers/feedback/good.py:67]:
```suggestion
        "😥 Thank you for your feedback, <@{details['user']}>!",
```

File: api/ask_astro/slack/controllers/feedback/good.py [hunk 10]
```diff
@@ -10,54 +12,81 @@
...
+    except SlackApiError as e:
+        # ignore the error if the reaction already exists
+        if e.response["error"] != "already_reacted":
+            raise e
+
+    try:
+        await client.reactions_remove(name="thumbsdown", channel=details["channel"], timestamp=details["message_ts"])
+    except Exception as e:
+        logger.debug("Failed to remove thumbsdown reaction: %s", e)
+        pass
```
Comment (Lee-W) [inline api/ask_astro/slack/controllers/feedback/good.py:92]:
```suggestion
```

File: api/ask_astro/slack/controllers/mention.py [hunk 11]
```diff
@@ -15,16 +16,76 @@
...
+    :param timestamp: Timestamp of the message.
+    """
+    try:
+        await client.reactions_add(
+            name=name,
+            channel=channel_id,
+            timestamp=timestamp,
+        )
+    except SlackApiError as exc:
+        logger.warning(f"Failed to add {name} reaction", exc_info=exc)
```
Comment (Lee-W) [inline api/ask_astro/slack/controllers/mention.py:40]:
```suggestion
        logger.warning("Failed to add %s reaction", name, exc_info=exc)
```

File: api/ask_astro/slack/controllers/mention.py [hunk 12]
```diff
@@ -34,116 +95,35 @@ async def on_mention(
         logger.error(f"Missing key: {e}")
         return
 
-    logger.info("Received question (%s): %s", ts, text)
+    logger.info(f"Received question ({ts}): {text}")
```
Comment (Lee-W) [inline api/ask_astro/slack/controllers/mention.py:98]: I think we'd better consolidate how we use logger

File: api/ask_astro/config.py (hunk 3 above)
Comment (sunank200) [reply api/ask_astro/config.py:54]: Yes OpenAI requires that

File: api/ask_astro/rest/controllers/get_request.py [hunk 13]
```diff
@@ -15,15 +16,23 @@
...
 
-    if not request.exists:
-        return json({"error": "Question not found"}, status=404)
+        if not request.exists:
+            return json({"error": "Question not found"}, status=404)
 
-    return json(request.to_dict(), status=200)
+        return json(request.to_dict(), status=200)
+    except Exception as e:
+        logger.error("Error fetching data for request %s: %s", request_id, str(e))
```
[the remark this answers is not in the document]
Comment (sunank200) [reply api/ask_astro/rest/controllers/get_request.py:37]: removed

File: api/ask_astro/rest/controllers/submit_feedback.py (hunk 5 above)
Comment (sunank200) [reply api/ask_astro/rest/controllers/submit_feedback.py:40]: removed

File: api/ask_astro/services/feedback.py (hunk 6 above)
Comment (sunank200) [reply api/ask_astro/services/feedback.py:60]: removed

File: api/ask_astro/rest/controllers/list_recent_requests.py (hunk 4 above)
Comment (sunank200) [reply api/ask_astro/rest/controllers/list_recent_requests.py:52]: removed

File: api/ask_astro/slack/controllers/feedback/bad.py (hunk 7 above)
Comment (sunank200) [reply api/ask_astro/slack/controllers/feedback/bad.py:67]: changed

File: api/ask_astro/slack/controllers/feedback/bad.py (hunk 8 above)
Comment (sunank200) [reply api/ask_astro/slack/controllers/feedback/bad.py:92]: We pass it as there is already a thumbs up

File: api/ask_astro/slack/controllers/feedback/good.py (hunk 9 above)
Comment (sunank200) [reply api/ask_astro/slack/controllers/feedback/good.py:67]: changed

File: api/ask_astro/slack/controllers/feedback/good.py (hunk 10 above)
Comment (sunank200) [reply api/ask_astro/slack/controllers/feedback/good.py:92]: removed it

File: api/ask_astro/slack/controllers/feedback/bad.py (hunk 8 above)
Comment (sunank200) [reply api/ask_astro/slack/controllers/feedback/bad.py:92]: removed it though

File: api/ask_astro/slack/controllers/mention.py (hunk 11 above)
Comment (sunank200) [reply api/ask_astro/slack/controllers/mention.py:40]: Changed. I missed this

File: api/ask_astro/slack/controllers/mention.py (hunk 12 above)
Comment (sunank200) [reply api/ask_astro/slack/controllers/mention.py:98]: Changed it

File: .gitignore [hunk 14]
```diff
@@ -245,3 +245,8 @@ airflow/include/data/astronomer/registry/registry_cells.parquet
 airflow/include/data/astronomer/registry/registry_dags.parquet
 airflow/include/data/baseline_data_*.parquet
 airflow/include/data/slack/troubleshooting.parquet
+<<<<<<< Updated upstream
+/api/gcp.json
+=======
+api/gcp.json
+>>>>>>> Stashed changes
```
Comment (pankajastro) [inline .gitignore:252]: this does not look right https://github.com/astronomer/ask-astro/pull/71

**Pull request merged**